capsules/src/virtual_aes_ccm.rs

```diff
@@ -0,0 +1,798 @@
+//! aes_cbc:    \__________________________________________________________/
+//! aes_ctr:                        \ 1 blk | _____________________________/
+//! ```
+//!
+//! At decryption, there is no choice but the reverse the order of operations.
+//! First, we zero out the overlapping block and perform ctr over it and the
+//! PlaintextData. This produces Enc(Key, A_i), which we save in saved_tag.
+//! Then, we restore the previous value of the last block of AuthData and re-pad
+//! PlaintextData before running CBC over both fields. The last step is to
+//! combine saved_tag and the unencrypted tag to form the encrypted tag and
+//! verify its correctness.
+//!
+//! Usage
+//! -----
+//!
+//! ```rust
+//! # use capsules::test::aes_ccm::Test;
+//! # use capsules::virtual_aes_ccm;
+//! # use kernel::hil::symmetric_encryption::{AES128, AES128CCM, AES128_BLOCK_SIZE};
+//! # use kernel::static_init;
+//! # use sam4l::aes::{Aes, AES};
+//! type AESCCMMUX = virtual_aes_ccm::MuxAES128CCM<'static, Aes<'static>>;
+//! type AESCCMCLIENT = virtual_aes_ccm::VirtualAES128CCM<'static, AESCCMMUX>;
+//! // mux
+//! let ccm_mux = static_init!(AESCCMMUX, virtual_aes_ccm::MuxAES128CCM::new(&AES));
+//! AES.set_client(ccm_mux);
+//! const CRYPT_SIZE: usize = 7 * AES128_BLOCK_SIZE;
+//! let crypt_buf1 = static_init!([u8; CRYPT_SIZE], [0x00; CRYPT_SIZE]);
+//! let ccm_client1 = static_init!(
+//!     AESCCMCLIENT,
+//!     virtual_aes_ccm::VirtualAES128CCM::new(ccm_mux, crypt_buf1)
+//! );
+//! ccm_client1.setup();
+//! let data1 = static_init!([u8; 4 * AES128_BLOCK_SIZE], [0x00; 4 * AES128_BLOCK_SIZE]);
+//! let t1 = static_init!(Test<'static, AESCCMCLIENT>, Test::new(ccm_client1, data1));
+//! ccm_client1.set_client(t1);
+//! let crypt_buf2 = static_init!([u8; CRYPT_SIZE], [0x00; CRYPT_SIZE]);
+//! let ccm_client2 = static_init!(
+//!     AESCCMCLIENT,
+//!     virtual_aes_ccm::VirtualAES128CCM::new(ccm_mux, crypt_buf2)
+//! );
+//! ccm_client2.setup();
+//! let data2 = static_init!([u8; 4 * AES128_BLOCK_SIZE], [0x00; 4 * AES128_BLOCK_SIZE]);
+//! let t2 = static_init!(Test<'static, AESCCMCLIENT>, Test::new(ccm_client2, data2));
+//! ccm_client2.set_client(t2);
+//! t1.run();
+//! t2.run();
+//!
+//! ```
+
+use crate::net::stream::SResult;
+use crate::net::stream::{encode_bytes, encode_u16};
+use core::cell::Cell;
+use kernel::common::cells::{OptionalCell, TakeCell};
+use kernel::common::{List, ListLink, ListNode};
+// use kernel::debug;
+use kernel::hil::symmetric_encryption;
+use kernel::hil::symmetric_encryption::{
+    AES128Ctr, AES128, AES128CBC, AES128_BLOCK_SIZE, AES128_KEY_SIZE, CCM_NONCE_LENGTH,
+};
+use kernel::ReturnCode;
+
+// the CCM state is not public is aes_ccm.rs
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+enum CCMState {
+    Idle,
+    Auth,
+    Encrypt,
+}
+
+// to cache up the function parameters of the crypt() function
+struct CryptFunctionParameters {
+    buf: &'static mut [u8],
+    a_off: usize,
```

Comment:
please explicit variable names, `a_off` is hard to understand I think `authdata_offset` is better or something similar.

Reply:
Ah sorry, it was originally called a_off in aes_ccm.rs, I don't know what they exactly mean TBH... sorry...

Reply:
So do you want me to expand those names? I am not very sure about them and there's a risk of mis-interpretation... Sorry for this... The code is from aes_ccm.rs.

Reply:
Let's hear what the others think but if no one understands what they mean that's a problem :)

Could be maybe fixed in another PR then.

Reply:
Thanks Niklas!